packages/data-stores/src/purchases/queries/use-site-purchases.ts

```diff
@@ -0,0 +1,44 @@
+import { useQuery, type UseQueryResult } from '@tanstack/react-query';
+import wpcomRequest from 'wpcom-proxy-request';
+import { createPurchaseObject } from '../lib/assembler';
+import useQueryKeysFactory from './lib/use-query-keys-factory';
+import type { RawPurchase, Purchase } from '../types';
+
+interface PurchasesIndex {
+	[ purchaseId: number ]: Purchase;
+}
+
+interface Props {
+	siteId?: string | number | null;
+}
+
+/**
+ * Fetches all purchases for a given site, transformed into a map of purchaseId => Purchase
+ * @param {Object} props - The properties for the function
+ * @param props.siteId Site ID
+ * @returns Query result
+ */
+function useSitePurchases( { siteId }: Props ): UseQueryResult< PurchasesIndex > {
+	const queryKeys = useQueryKeysFactory();
+
+	return useQuery( {
+		queryKey: queryKeys.sitePurchases( siteId ),
+		queryFn: async (): Promise< PurchasesIndex > => {
+			const purchases: RawPurchase[] = await wpcomRequest( {
+				path: `/sites/${ encodeURIComponent( siteId as string ) }/purchases`,
+				apiVersion: '1.1',
+			} );
+
+			return Object.fromEntries(
+				purchases.map( ( rawPurchase ) => {
+					const purchase = createPurchaseObject( rawPurchase );
+					return [ purchase.id, purchase ];
+				} )
+			);
+		},
+		staleTime: 1000 * 60 * 5, // 5 minutes
```

Comment:
I'm curious: why this 5 minute delay?

I know that right now this is only used for the pricing grid, but this hook seems pretty general-purpose to me. I'm imagining the case where a user makes a purchase, completes checkout, and then goes to view their purchase. In that case, a five minute delay before it appears seems like a poor experience.

Reply:
Thanks fof commenting. I imagine that by the time this hook is used at other places in the app, the rest of Purchases state will migrate here, including mutations, which would force these queries to be invalidated and refetched. That said, mutations can also come from elsewhere or in some uncontrolled fashion. I'm also imagining of an imaginary friend from @Automattic/payments to own the rest of the migration of this state slice with its mutations etc. to data stores. I guess that's all a little convoluted.

What value do you suggest to use?

Reply:
Hm... you're right. It's true that really we should force flows which update purchases to invalidate the cache but on the other hand, I have to ask myself: what's the reason for a delay at all?

Without this, the hook will refetch the purchases every time the page loads, or if the page is refocused after a certain period of time has passed. I'm guessing this line was added so that the cache lasts longer than a page refresh. But why? Do you expect that users will frequently be reloading the plans grid multiple times within 5 minutes? Is the endpoint particularly slow or expensive? Or is the loading UI particularly disruptive?

If setting a `staleTime` was just trying to prevent a problem that hasn't happened yet, then I'd say it's premature optimization. If there _is_ a reason, then we probably do want to put the requirement on the consumer to perform invalidation like you suggested.

Reply:
> Without this, the hook will refetch the purchases every time the page loads, or if the page is refocused after a certain period of time has passed

This sounds 100% fine to me. I think the intention of this line was more to "not refetch within 5 minutes of interacting with the grid", as we've done with the rest.

Reply:
Since 0 is the default, it lets the query provider use its default logic which is what we use on most calls to `useQuery`. You can learn a little about its default behavior here: https://tkdodo.eu/blog/practical-react-query#the-defaults-explained or here: https://tanstack.com/query/v4/docs/react/guides/ssr#staleness-is-measured-from-when-the-query-was-fetched-on-the-server

Basically, it will refetch on page load and occasionally on refocus or if the network reconnects. AFAIK it will never refetch while the page is loaded and focused, no matter how many times it is re-rendered or what sort of interaction happens unless something explicitly invalidates the cache. In my experience these defaults usually work just fine in most circumstances.

The main reason to change `staleTime` is to prevent fetching when the page reloads, and based on what you said I don't think there's any reason to do that here.

Reply:
I don't think you need to worry about unnecessary processing of the component in case a fetch happens. We have to consider that when we as devs work on those queries our behaviour is vastly different from how users interact with the app and I don't think fetching too often is a problem for users most of the time. It won't cause a rerender if the data didn't change (although that's on the React side of things).

> The main reason to change staleTime is to prevent fetching when the page reloads

I don't think that's the main reason even though it's one of the effects. The [first article you shared](https://tkdodo.eu/blog/practical-react-query#the-defaults-explained) explains how it works pretty well and IMO the `staleTime` value should be not set by default or relatively low for custom query hooks but it highly depends on the data. 

I don't think you should ever set `staleTime` to `Infinity` by default. If there's really a need for a high `staleTime` you should make it possible to set via `queryOptions` on your custom query hook and not affect other usages of your hook (which may need fresh data).

If we're comparing with our Redux data layer data will be fetched on a page reload, as a user I would expect that behaviour and I think we should keep it like that.

I know sometimes devs deactivate `refetchOnWindowFocus` or set a high `staleTime` to prevent seemingly unnecessary fetches but I think by doing that they're preventing where RQ really shines.

Reply:
> It won't cause a rerender if the data didn't change (although that's on the React side of things).

I think it would cause a rerender at least from the various flags (`isFetching`, etc.) being reset, depending on how/if one binds the UI around those to show a loader or something (and we currently do that). 

> If we're comparing with our Redux data layer data will be fetched on a page reload, as a user I would expect that behaviour and I think we should keep it like that.

I think the default `staleTime` being 0 and `refetchOnWindowFocus` being false is an aggressive strategy. We are assuming that unfocusing the window and refocusing is ok to force the entire data layer be recollected and recomputed. That will surely lead to loaders and potentially flickering happening - in a React context, it sounds like the entirety of the app may rerender in that case if this model is applied holistically and uniformly. Hence the framework exposes these properties for tuning. In my experience, viewing a site and interacting is only part of what I do when I'm on my phone or tablet or laptop (I'd find it hilarious if I click out and into a products website and see a bunch of loaders go off on the screen every time)

I need to investigate a little more, but my first impression - we are not serving the user experience 100% with this strategy (if the behavior is what I outline above).

Reply:
> I don't think you should ever set staleTime to Infinity by default.

Yeah, agreed. I don't think setting to infinity by default would be a good way to go of course. I am only outlining when that would make sense - like a spectrum type of thing (the two ends).

>  If there's really a need for a high staleTime you should make it possible to set via queryOptions on your custom query hook and not affect other usages of your hook (which may need fresh data).

This might be something worth exploring, however, not sure how that would be factored exactly. I think for the moment, default options and reworking the UI to not depend on these data being in a fetching state would be the way forward. I need to confirm a few things as we have whole page loaders dependent on `isFetching` (which are being refactored elsewhere - so that may need to happen first)

Reply:
> It won't cause a rerender if the data didn't change (although that's on the React side of things).

and I think in the current form (transforming inside the `queryFn`) the data changing will always be the case, as a transformation in `queryFn` would be recomputed on every fetch. So anything that consumes that data would be getting a new reference, no?

Reply:
I think we may be mixing different things. One the one side we have RQ which may refetch on certain conditions but that doesn't mean the UI layer (React) does update anything. If a list of items is still the same, it won't update and nothing is going to flicker.

> I'd find it hilarious if I click out and into a products website and see a bunch of loaders go off on the screen every time

There is different fetch states which you might be subscribing to like `isLoading`, `isFetching`, etc – as you mentioned. If you base UI actions around those it's expected those result in some kind of signal. A background refetch though doesn't have to result in any UI loaders going off. Usually, what you're subscribing to is `isInitialLoading` (or `isLoading` in v5) to indicate the first fetch of a query (i.e. we don't have any data yet) but it doesn't change it's state on subsequent fetches. I think this distinction is important.

Reply:
> This is the effect right now with the loader/spinner based on the isFetching status of this and the plans queries.

I only saw [your last comment](https://github.com/Automattic/wp-calypso/pull/84761#discussion_r1423980027) after I submit mine but it may be that you don't want to use `isFetching` (whenever the queryFn is executing) but `isInitialLoading` (whenever the first fetch of a query is running).

Reply:
Sorry for derailing. I'll rework things (indeed looking into `isLoading` now). The intention with `isFetching` was to never have updated pricing being injected into the grid without some form of notice (the spinner - which was previously a blank page). The case stems from having multiple sources right now providing these data (multiple query hooks for plans & Calypso state & next this Purchases hook -> they are all feeding into the final price display). 

You can ignore my comments. If it doesn't work out, we'll address in a follow-up.

Reply:
> If a list of items is still the same, it won't update and nothing is going to flicker.

@flootr  Interesting. If the `data` reference changes (and it would if it's transformed in the `queryFn`, no?), wouldn't that cause a cascading effect on anything that consumes it - to either rerender or recompute some transformation from it?

Reply:
If the `data` reference changes it will definitely do all of that, but React won't apply changes to the DOM if there's nothing to update. This is what I meant when I said we may speak of different things.

I created a small sandbox to demonstrate: https://codesandbox.io/p/sandbox/young-glade-z79ygv – you can see that the first list of items doesn't restart the animation after you hit the _Refetch_ button. It does for the second list of items because the order changed.

Reply:
You are right. I guess if your display logic depends on a certain transformation having settled, then would force some UI change, but yeah, probably not flickering if leaf nodes don't change.